spec/integration/application/agent_spec.rb

```diff
@@ -633,4 +633,35 @@
       end
     end
   end
+
+  context "environment convergence" do
+    it "switches to 'newenv' environment and retries the run" do
+      first = true
+
+      catalog_handler = -> (req, res) {
+        code = if first
+                 first = false
+                 ''
+               else
+                 'notify { "converged": }'
+               end
+
+        catalog = compile_to_catalog(code)
+        catalog.environment = 'newenv'
+
+        res.body = formatter.render(catalog)
+        res['Content-Type'] = formatter.mime
+      }
+
+      server.start_server(mounts: {catalog: catalog_handler}) do |port|
+        Puppet[:serverport] = port
+        expect {
+          agent.command_line.args << '--test' << '--debug'
+          agent.run
+        }.to exit_with(2)
+         .and output(a_string_matching(%r{Notice: Local environment: 'production' doesn't match server specified environment 'newenv', restarting agent run with environment 'newenv'})
+         .and matching(%r{Notice: converged})).to_stdout
```

Comment:
This test would also pass without the new code, is it worth it to have a test that checks if an empty catalog is returned by the compiler?

Reply:
Good point, I’ll take a look

Reply:
I updated the test to be more realistic, so it covers the scenario where a fact doesn't exist in the first environment, but does in the second, and is needed for compilation to succeed. But the actual short-circuit logic can't be tested in the integration test, since it's server side logic that runs in puppetserver. So I updated the `spec/unit/indirector/catalog/compiler_spec.rb` to cover the default behavior and the behavior when `check_environment` is set.